### PR TITLE
Profiler docstrings

### LINKED ISSUE #397: Docstrings cleanup! `composer.profiler`
The `composer.profiler` needs docstrings. As you work through adding docstrings, please:

## Things to pay attention to
1. Ensure that each public module has a module-level docstring.
   i. _Need to make a module private? Rename it so the file (or folder) begins with an underscore._ 
2. On each public module (e.g. `__init__.py`, as well as individual public files), ensure that `__all__` is set to the members that should be re-exported. This is necessary so our API docs will properly list the members of that file, and not all imported members (such as from python's stdlib imports).
2. Ensure that each public class has a class-level docstring
  i. _Need to make a class private? Rename it so the class name begins with an underscore._
3. Ensure that each public function (except for `__init__`, which is handled by the class-level docstring) has a docstring.
  i. _Need to make a function private? Rename it so the function name begins with an underscore._
4. Double-check cross references. Double-tick marks (e.g. ``` ``code`` ```) should ONLY be used for values or function parameter names; never for attributes, functions, classes, or modules. We have [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) configured, so this applies to imports as well. Instead, for a cross reference, use the following format:
  i. `` :attr:`MyClass.attribute` `` for attributes
  i. `` :meth:`MyClass.my_method` `` for methods
  i `` :class:`MyClass` `` for classes
  i: `` :mod:`composer.utils` `` for modules
5. Any public functions that we expect a user to invoke should have an example. Please use [doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) to structure the example codes. Doctest ensures that the examples actually run and works and do not get out-of-date.

## Previewing docs locally

```bash
cd docs
python3 -m http.server --directory _build/html/ &  # run a web server on http://localhost:8000
rm -rf source/api_reference && make clean && make html  # clean the api reference docstrings, and rebuild the docs folder
```

Then, go to http://localhost:8000/api_reference.html to preview the api reference

### PR DESCRIPTION
**Pull request by bandish-shah**
Resolves #397

### CODE REVIEW

**Comment by ravi-mosaicml** [comment]
I took a first pass through the preview of the docs build (didn't look at any code changes).

While this is a good start, I think this PR could benefit from a second pass through it. All public modules will need `__all__` to be defined along with a module-level docstring. In addition, it looks like there are still some incomplete docstrings and others that are misformatting. I think we should also think about what classes and/or modules are public vs what should be private and not part of the user-facing API.

Below are some specific comments:

- [ ] For https://mosaicml-composer--473.com.readthedocs.build/en/473/api_reference/composer.profiler.html#module-composer.profiler:
  - [ ] The formatting under "profilers include" seems off -- specifically where line breaks are placed and how  subsequent lines for each bullet are indented.
  - [ ] Can you include an example of what the user needs to do to use the profiler? Probably an example with the trainer init?
  - [ ] Can we include a screenshot of the trace and a tutorial on how to visualize the trace?
- [ ] Would it make sense to make the torch profiler, dataloader profiler, and system profiler modules private, and to not re-export the underlying callbacks? Since users are no longer constructing these classes directly, I don't think we need to expose the underlying implementation of the profiler in the docs (also fewer docstrings to write lol)
- [ ] https://mosaicml-composer--473.com.readthedocs.build/en/473/api_reference/composer.profiler.event_handler.html#module-composer.profiler.event_handler
  - [ ] Missing a module level docstring.
  - [ ] In addition, the __init__ docstring doesn't mention `process_counter_event`
  - [ ] References are not auto-linking
 - [ ] https://mosaicml-composer--473.com.readthedocs.build/en/473/api_reference/composer.profiler.json_trace.html#module-composer.profiler.json_trace could use an example on how to visualize the JSON format
 - [ ] https://mosaicml-composer--473.com.readthedocs.build/en/473/api_reference/composer.profiler.profiler_hparams.html#module-composer.profiler.profiler_hparams: Can you add a YAML example for the profiler hparams?

**Review by ravi-mosaicml — COMMENTED** [review]
Left feedback offline

File: composer/profiler/torch_profiler.py [hunk 1]
```diff
@@ -31,7 +42,7 @@ class TorchProfiler(Callback):
         pip install tensorbaord torch_tb_profiler
         tensorboard --logdir tensorboard_trace_handler_dir
 
-    Also see https://pytorch.org/docs/stable/profiler.html.
+    Also see :doc:`profiler`.
```
Comment (ravi-mosaicml) [inline composer/profiler/torch_profiler.py:45]: Missing a "the" in the note below...it should read...

When record_shapes is specified, THE profiler will temporarily hold...

Comment (ravi-mosaicml) [reply composer/profiler/torch_profiler.py:45]: Also docstring below is missing a tick mark around torch_profiler

File: composer/profiler/_profiler.py [hunk 2]
```diff
@@ -320,7 +350,17 @@ def __init__(self, profiler: Profiler, name: str, actions: Sequence[ProfilerActi
         self._action_at_start = None
 
     def start(self) -> None:
-        """Record the start of a duration event."""
+        """Record the start of a duration event.
+
+        To record the duration of an event, invoke :meth:`Marker.start` followed by :meth:`Marker.finish`\\:
+
+        .. code-block:: python
```
Comment (ravi-mosaicml) [inline composer/profiler/_profiler.py:357]: doctest

File: composer/profiler/_profiler.py [hunk 3]
```diff
@@ -374,10 +417,19 @@ def finish(self) -> None:
         self._started = False
 
     def instant(self) -> None:
-        """Record an instant event."""
+        """Record an instant event.
+
+        To record an instant event:
+
+        .. code-block:: python
```
Comment (ravi-mosaicml) [inline composer/profiler/_profiler.py:424]: doctest

File: composer/profiler/_profiler.py [hunk 4]
```diff
@@ -386,10 +438,21 @@ def instant(self) -> None:
             )
 
     def counter(self, values: Dict[str, Union[float, int]]) -> None:
-        """Record a counter event."""
+        """Record a counter event.
+
+        To record a counter event:
+
+        .. code-block:: python
```
Comment (ravi-mosaicml) [inline composer/profiler/_profiler.py:445]: doctest

File: composer/profiler/_profiler.py [hunk 5]
```diff
@@ -9,43 +9,46 @@
...
-__all__ = ["Profiler", "Marker"]
+log = logging.getLogger(__name__)
 
 
 class Profiler:
-    """The Profiler produces a trace of the training graph.
+    """Records the duration of Trainer :class:`.Event` using the :class:`.Marker` API.
 
-    Specifically, it records:
+    The :class:`Profiler` is instantiated by the Composer :class:`.Trainer`\\.
```
Comment (ravi-mosaicml) [inline composer/profiler/_profiler.py:29]: `..note ::` notation?

File: composer/profiler/json_trace.py [hunk 6]
```diff
@@ -16,10 +18,15 @@
...
+
 
 class JSONTraceHandler(ProfilerEventHandler):
-    """Records trace events in `JSON trace format <https://\\
-    docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview>`_.
+    """Records trace events in `JSON trace format
+    <https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview>`_.
+
+    Traces are output to ``output_directory``.  Traces can be visualized using the Chrome Trace Viewer.
+    To view in a Google Chrome browser, navigate to ``chrome://tracing`` and load the JSON trace file.
```
Comment (ravi-mosaicml) [inline composer/profiler/json_trace.py:29]: Can chrome://tracing be in link format?

File: composer/profiler/__init__.py [hunk 7]
```diff
@@ -1,5 +1,52 @@
...
+
+It is required to specify an output ``profiler_trace_file`` during :class:`.Trainer` initialization to enable profiling.
+The ``profiler_trace_file`` will contain the profiling trace data once the profiling run completes.  By default, the :class:`.Profiler`,
+:class:`.DataloaderProfiler` and :class:`.SystemProfiler` will be active.  The :class:`.TorchProfiler` is **disabled** by default.
+
+To activate the :class:`.TorchProfiler`, the ``torch_profiler_trace_dir`` must be specified *in addition* to the ``profiler_trace_file`` argument.
+The ``torch_profiler_trace_dir`` will contain the Torch Profiler traces once the profiling run completes.  The :class:`.Profiler` will
+automatically merge the Torch traces in the ``torch_profiler_trace_dir`` into the ``profiler_trace_file``, allowing users to view a unified trace.
+
+The complete traces can be viewed by in a Google Chrome browser navigating to ``chrome://tracing`` and loading the ``profiler_trace_file``.
```
Comment (ravi-mosaicml) [inline composer/profiler/__init__.py:28]: Can chrome://tracing be in link format?

File: composer/profiler/json_trace_merger.py [hunk 8]
```diff
@@ -59,16 +61,13 @@ def _get_rank_to_clock_syncs(trace_files: Tuple[str, ...]) -> Dict[int, int]:
...
-    Each trace file must contain ``global_rank`` in the metadata, and
-    one trace file per rank must contain ``clock_sync_timestamp_us`` in the metadata.
+    """Merge profiler output JSON trace files together.
 
     This function will update the trace events such that:
 
     - The ``pid`` will be set to the global rank.
     - The ``ts`` is syncronized with that of the rank 0 process.
-    - Ensures that the backward process is one below the forward process
+    - Ensures that the backward process appears below the forward process
```
Comment (ravi-mosaicml) [inline composer/profiler/json_trace_merger.py:70]:
```suggestion
    - The backward pass process appears below the forward process.
```

**Review by ravi-mosaicml — COMMENTED** [review]
Last round of comments; all nits. Looks awesome!!

File: composer/profiler/__init__.py (hunk 7 above)
Comment (bandish-shah) [reply composer/profiler/__init__.py:28]: No I originally tried this but it doesn't actually open the trace viewer via a hyperlink.  Not sure if there's some magic to enable which is why I explicitly state how to open the trace viewer in all relevant areas.

File: composer/profiler/json_trace.py (hunk 6 above)
Comment (bandish-shah) [reply composer/profiler/json_trace.py:29]: Nope

**Review by ravi-mosaicml — APPROVED** [review]
:rocket:

**Pull request merged**